### PR TITLE
add object to represent the Native Window

### PR DESCRIPTION
**Pull request by RemiDufloer**
I added an object to represent the Native Window,  and i added a manager for them.

The following code is used to open a native window with its manager :

```st
manager := MorphicWindowManager new.
p := manager newWindowWithRoot: Morph new.
p open.
```

or this :

```st
manager := MorphicWindowManager new.
newWindow := manager openNewWindowWithRoot: ClyFullBrowserMorph onDefaultEnvironment.
```

and to close :

```st
manager ensureDeleteAll .
```

It allows to obtain this kind of result :

Capture d’écran 2023-04-27 à 11 45 09

### CODE REVIEW

**Review by jecisc — APPROVED** [review]
Oh nice!

The feature is cool and the code is nice with documentation and tests. 
I'm not sure what are the plan for native windows so maybe someone else could review too. But if it goes in the aimed direction this is a great addition!

Maybe we could also add to the windows menu (the little arrow at the corner of the window) an option "Open natively" to make it easier to use and more discoverable

**Comment by Ducasse** [comment]
Thanks Remi!!! This is cool that you push this. We got distracted and could not deliver your cool code for P11.

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+	newWindow window: aMorph.
+	newWindow manager: aManager.
+	^ newWindow 
+	
+]
+
+{ #category : #closing }
+MorphicNativeWindow >> allowedToClose [
+
+" no reason to dont close ? "
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:47]: For following iteration: Fix comment.
Also, why not allowed to close?

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+]
+
+{ #category : #'submorphs - add/remove' }
+MorphicNativeWindow >> attributs: aOsWindowAttributs [
+
+	attributes := aOsWindowAttributs
+]
+
+{ #category : #'submorphs - add/remove' }
+MorphicNativeWindow >> attributsSave [
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:64]: For following iteration: 
  attributs => attributes

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+
+	osWindow position.
+	osWindow title.
+	osWindow bounds.
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:68]: For following iteration: 
Calling these methods are saving the attributes?
Add a comment!

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+MorphicNativeWindow >> backend [
+
+	^ backend
+]
+
+{ #category : #testing }
+MorphicNativeWindow >> canBeMaximized [
+	"Answer whether we are not we can be maximised."
+
+	^ attributes  resizable  and: [ attributes maximized not ]
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:83]: For following iteration: reformat please :)

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+	Use #delete instead"
+
+	self isOpen ifFalse: [ self error: 'Cannot delete non open window' ].
+	self ensureDoDelete
+]
+
+{ #category : #'submorphs - add/remove' }
+MorphicNativeWindow >> ensureDelete [
+
+	self ensureDoDelete.
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:126]: For following iteration:

could we add a comment of why this method? What's the difference with doDelete? when should I use one or the other?

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+MorphicNativeWindow >> isNotMinimized [
+
+	^ attributes minimized not
+]
+
+{ #category : #testing }
+MorphicNativeWindow >> isNotRestored [
+	"Answer whether we are maximised or minimised."
+
+	^attributes  minimized or: [attributes maximized]
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:176]: For following iteration: reformat

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+{ #category : #'window management' }
+MorphicNativeWindow >> restore [
+
+	osWindow restore
+]
+
+{ #category : #'taskbar-required' }
+MorphicNativeWindow >> taskbarButtonIsActive [
+
+	"true/false"
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:240]: For following iteration:

Pretty useless comment :)

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+		icon: theme windowCloseForm;
+		font: theme menuFont.
+
+	^menu
+]
+
+{ #category : #icons }
+MorphicNativeWindow >> taskbarIcon [
+
+ ^ window class taskbarIcon
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:329]: For following iteration: format

File: src/Morphic-Base/MorphicNativeWindow.class.st [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+]
+
+{ #category : #'taskbar-required' }
+MorphicNativeWindow >> taskbarLabel [
+	"Answer the label to use for a taskbar button for the receiver."
+	
+	^ self attributs title
```
Comment (guillep) [inline src/Morphic-Base/MorphicNativeWindow.class.st:336]: For following iteration:

typo attributs => attributes

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 10]
```diff
@@ -0,0 +1,200 @@
+"
+I manage the native window, when i was create i open a window with a taskbar, if you create a window, the window is add at the taskbar.
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: For following iteration:

When I am created I open a window with a taskbar.

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+		'rootSave',
+		'allAttributes',
+		'windows'
+	],
+	#category : #'Morphic-Base'
+}
+
+{ #category : #accessing }
+MorphicWindowManager >> allAttributes [
+	^ allAttributes
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st:42]: For following iteration: format

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+	(windows at: m) delete ].
+	taskBar updateTasks
+]
+
+{ #category : #accessing }
+MorphicWindowManager >> closeAllWindows [
+
+	
+	(windows size) to: (1) by: (-1) do: [ :m | (windows at: m) delete  ].
+	taskBar updateTasks .
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: For following iteration: format

Comment (guillep) [reply src/Morphic-Base/MorphicWindowManager.class.st]: Why are windows deleted in reverse order?
That could be added as a comment

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+]
+
+{ #category : #accessing }
+MorphicWindowManager >> closeWindow: aWindow [
+
+	(windows includes: aWindow) ifTrue: [ windows remove: aWindow ].
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st:87]: Could it happen that we are asked to close a window that we don't have?

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+]
+
+{ #category : #closing }
+MorphicWindowManager >> deleteAll [
+
+	self closeAllWindows.
+
+	"Call doDelete because the taskbar window is not subscribed to manager"
+	taskBarWindow doDelete.
+
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: For following iteration: format

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+	taskBar updateTasks
+]
+
+{ #category : #closing }
+MorphicWindowManager >> ensureDeleteAll [
+
+	self ensureCloseAllWindow.
+
+	"Call doDelete because the taskbar window is not subscribed to manager"
+	taskBarWindow ensureDelete
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: Comment != code :)

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+	"Call doDelete because the taskbar window is not subscribed to manager"
+	taskBarWindow doDelete.
+
+]
+
+{ #category : #accessing }
+MorphicWindowManager >> ensureCloseAllWindow [
+
+	windows size to: 1 by: -1 do: [ :m | 
+		(windows at: m) ensureDelete ].
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: Here it looks repeated with closeAllWindows.

Also this method should be called "ensureCloseAllWindows"

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+	"Call doDelete because the taskbar window is not subscribed to manager"
+	taskBarWindow ensureDelete 
+]
+
+{ #category : #accessing }
+MorphicWindowManager >> initialize [
+
+	windows := OrderedCollection new.
+	self startupTaskbar .
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st]: For following iteration: format

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+	^ windows size
+]
+
+{ #category : #accessing }
+MorphicWindowManager >> openNewWindowWithRoot: aMorph [ 
+
+	| newWindow |
+	newWindow := ( self newWindowWithRoot: aMorph ) open.
+	taskBar updateTasks .
+	^ newWindow
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st:182]: For following iteration: format

File: src/Morphic-Base/MorphicWindowManager.class.st [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+MorphicWindowManager >> saveAllRoot [
+	
+	rootSave := windows collect: [ :m | m window ]
+]
+
+{ #category : #startup }
+MorphicWindowManager >> startupOpenWithAttributes [
+
+	1 to: allAttributes size do: [ :a | 
+	(windows at: a) openWithAttributes: (allAttributes at: a) ]
```
Comment (guillep) [inline src/Morphic-Base/MorphicWindowManager.class.st:201]: For following iteration: format

File: src/Morphic-Tests/AbstractWindowManagerTest.class.st [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+	
+]
+
+{ #category : #tests }
+AbstractWindowManagerTest >> tearDown [ 
+
+	manager ensureDeleteAll.
+	super tearDown
+
+
```
Comment (guillep) [inline src/Morphic-Tests/AbstractWindowManagerTest.class.st:30]: For following iteration: format

File: src/Morphic-Tests/AbstractWindowManagerTest.class.st [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+		'manager'
+	],
+	#category : #'Morphic-Tests'
+}
+
+{ #category : #tests }
+AbstractWindowManagerTest >> runCaseManaged [
+
+	super runCaseManaged .
+	self runCase
```
Comment (guillep) [inline src/Morphic-Tests/AbstractWindowManagerTest.class.st:14]: We should keep only one of both

File: src/Morphic-Tests/MorphicNativeWindowTest.class.st [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+{ #category : #tests }
+MorphicNativeWindowTest >> testMinimizeWindow [
+
+	"window open.
+
+	window minimize.
+
+	self assert: window isMinimized.
+	
+	window restore"
```
Comment (guillep) [inline src/Morphic-Tests/MorphicNativeWindowTest.class.st]: This test is all commented

File: src/Morphic-Tests/MorphicNativeWindowTest.class.st [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+]
+
+{ #category : #tests }
+MorphicNativeWindowTest >> testNewWindowDefaultsIsNotMaximize [
+
+	window open .
+	
+	self assert: window isMaximized equals: false .
```
Comment (guillep) [inline src/Morphic-Tests/MorphicNativeWindowTest.class.st]: For following iteration: format

File: src/Morphic-Tests/MorphicNativeWindowTest.class.st [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+	self assert: window isNotMinimized
+]
+
+{ #category : #tests }
+MorphicNativeWindowTest >> testRestoreWindow [
+
+	window open .
+	window restore .
+	
+	self assert: window isNotRestored  equals: false  .
```
Comment (guillep) [inline src/Morphic-Tests/MorphicNativeWindowTest.class.st]: For following iteration: format

File: src/Morphic-Tests/MorphicWindowManagerTest.class.st [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+MorphicWindowManagerTest >> testWindowShouldBeSaveBeforeShutDownSaveAttributes [
+
+	manager openNewWindowWithRoot: Morph new.
+	manager saveAllAttributes.
+	
+	self assert: manager numberOfAttributesSave equals: 1.
+	
+
+	
+
```
Comment (guillep) [inline src/Morphic-Tests/MorphicWindowManagerTest.class.st]: For following iteration: format

File: src/Morphic-Widgets-Taskbar/TaskbarMorph.class.st [hunk 26]
```diff
@@ -428,7 +428,7 @@ TaskbarMorph >> updateTasks [
 	task will be considered dead along with a new replacement."
 
 	| wm deadTasks newTasks taskbarTasksFromWorldMorph |
-	wm := self worldMorphs asOrderedCollection.
+	wm := self windows.
```
Comment (guillep) [inline src/Morphic-Widgets-Taskbar/TaskbarMorph.class.st:431]: nice!

File: src/Morphic-Widgets-Taskbar/TaskbarMorphForManager.class.st [hunk 27]
```diff
@@ -0,0 +1,32 @@
+"
+is same of taskBarMorph but just for the manager
+"
+Class {
+	#name : #TaskbarMorphForManager,
+	#superclass : #TaskbarMorph,
```
Comment (guillep) [inline src/Morphic-Widgets-Taskbar/TaskbarMorphForManager.class.st:6]: Could these two classes be merged in the next interations?

Comment (guillep) [reply src/Morphic-Widgets-Taskbar/TaskbarMorphForManager.class.st:6]: Like, always having a window manager

File: src/OSWindow-Core/OSWindow.class.st [hunk 28]
```diff
@@ -177,9 +184,11 @@ OSWindow >> deliverGlobalEvent: aGlobalEvent [
 
 { #category : #private }
 OSWindow >> destroy [
+	announcer  announce: DeleteWindowByCrossAnnouncement .
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:187]: Why should we differentiate the cross event from the programmatic one?

File: src/OSWindow-Core/OSWindow.class.st [hunk 29]
```diff
@@ -189,19 +198,19 @@ OSWindow >> diagonalDPI [
 
 { #category : #accessing }
 OSWindow >> eventHandler [
-
+	
 	^ eventHandler
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:202]: For following iteration: format

File: src/OSWindow-Core/OSWindow.class.st [hunk 30]
```diff
@@ -189,19 +198,19 @@ OSWindow >> diagonalDPI [
...
 ]
 
 { #category : #accessing }
 OSWindow >> eventHandler: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:207]: For following iteration: format

File: src/OSWindow-Core/OSWindow.class.st [hunk 31]
```diff
@@ -212,7 +221,7 @@ OSWindow >> extent: newExtent [
 { #category : #'window management' }
 OSWindow >> focus [
 
-	self validHandle raise
+	self validHandle raise.
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:224]: For following iteration: format

File: src/OSWindow-Core/OSWindow.class.st [hunk 32]
```diff
@@ -444,18 +465,18 @@ OSWindow >> title: aString [
...
 
-	self validHandle toggleBorderOff
+	self validHandle toggleBorderOff.
 ]
 
 { #category : #'window management' }
 OSWindow >> toggleBorderOn [
 
-	self validHandle toggleBorderOn
+	self validHandle toggleBorderOn.
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:474]: For following iteration: format

File: src/OSWindow-Core/OSWindow.class.st [hunk 33]
```diff
@@ -444,18 +465,18 @@ OSWindow >> title: aString [
 { #category : #'window management' }
 OSWindow >> toggleBorderOff [
 
-	self validHandle toggleBorderOff
+	self validHandle toggleBorderOff.
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:468]: For following iteration: format

File: src/OSWindow-Core/OSWindow.class.st [hunk 34]
```diff
@@ -433,7 +454,7 @@ OSWindow >> stopTextInput [
 
 { #category : #accessing }
 OSWindow >> title [
-	^ self validHandle title
+	^ initialAttributes title: (self validHandle title)
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:457]: So at the end the initialAttributes do not hold only the initial attributes. They hold the current attributes of the window, right? Could we rename the variable?

File: src/OSWindow-Core/OSWindow.class.st [hunk 35]
```diff
@@ -444,18 +465,18 @@ OSWindow >> title: aString [
...
 OSWindow >> toggleBorderOn [
 
-	self validHandle toggleBorderOn
+	self validHandle toggleBorderOn.
 ]
 
 { #category : #events }
 OSWindow >> updateToNewResolution [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindow.class.st:479]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 36]
```diff
@@ -111,7 +111,7 @@ OSWindowAttributes class >> initialize [
 	DefaultBorderless := false.
 	DefaultResizable := true.
 	DefaultMaximized := false.
-	DefaultMinimized := false
+	DefaultMinimized := false.
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:114]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 37]
```diff
@@ -122,7 +122,7 @@ OSWindowAttributes class >> newWithNullDriver [
 { #category : #accessing }
 OSWindowAttributes >> applyTo: window [
 	window title: self title;
-		icon: self icon
+		icon: self icon.
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:125]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 38]
```diff
@@ -132,7 +132,7 @@ OSWindowAttributes >> borderless [
 
 { #category : #accessing }
 OSWindowAttributes >> borderless: anObject [
-	borderless := anObject
+	^ borderless := anObject
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:135]: Why this setter returns the value?

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 39]
```diff
@@ -142,8 +142,11 @@ OSWindowAttributes >> bounds [
 
 { #category : #accessing }
 OSWindowAttributes >> bounds: aRectangle [
-	position := aRectangle origin.
-	extent := aRectangle extent
+
+	self position: aRectangle origin.
+	extent := aRectangle extent.
+	self windowCentered: false.
+	^ aRectangle
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:149]: same

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 40]
```diff
@@ -161,30 +164,30 @@ OSWindowAttributes >> extent [
 
 { #category : #accessing }
 OSWindowAttributes >> extent: newExtent [
-	extent := newExtent
+	^ extent := newExtent
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:167]: same

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 41]
```diff
@@ -161,30 +164,30 @@ OSWindowAttributes >> extent [
...
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> fullscreen [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:172]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 42]
```diff
@@ -161,30 +164,30 @@ OSWindowAttributes >> extent [
...
 OSWindowAttributes >> fullscreen [
-
+	
 	^ fullscreen
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> fullscreen: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:178]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 43]
```diff
@@ -161,30 +164,30 @@ OSWindowAttributes >> extent [
...
 OSWindowAttributes >> fullscreen: anObject [
-
+	
 	fullscreen := anObject
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> glAttributes [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:184]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 44]
```diff
@@ -161,30 +164,30 @@ OSWindowAttributes >> extent [
...
 OSWindowAttributes >> glAttributes [
-
+	
 	^ glAttributes
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> glAttributes: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:190]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 45]
```diff
@@ -195,13 +198,13 @@ OSWindowAttributes >> height [
 
 { #category : #accessing }
 OSWindowAttributes >> icon [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:201]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 46]
```diff
@@ -195,13 +198,13 @@ OSWindowAttributes >> height [
...
 OSWindowAttributes >> icon [
-
+	
 	^ icon
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> icon: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:207]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 47]
```diff
@@ -247,18 +250,18 @@ OSWindowAttributes >> position [
 
 { #category : #accessing }
 OSWindowAttributes >> position: aPoint [
-	position := aPoint
+	^ position := aPoint
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:253]: why the setter returns a value?

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 48]
```diff
@@ -247,18 +250,18 @@ OSWindowAttributes >> position [
...
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> preferableDriver [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:258]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 49]
```diff
@@ -247,18 +250,18 @@ OSWindowAttributes >> position [
...
 OSWindowAttributes >> preferableDriver [
-
+	
 	^ preferableDriver
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> preferableDriver: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:264]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 50]
```diff
@@ -284,13 +287,13 @@ OSWindowAttributes >> screenId: newScreenId [
 
 { #category : #accessing }
 OSWindowAttributes >> title [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:290]: For following iteration: format

File: src/OSWindow-Core/OSWindowAttributes.class.st [hunk 51]
```diff
@@ -284,13 +287,13 @@ OSWindowAttributes >> screenId: newScreenId [
...
 OSWindowAttributes >> title [
-
+	
 	^ title
 ]
 
 { #category : #accessing }
 OSWindowAttributes >> title: anObject [
-
+
```
Comment (guillep) [inline src/OSWindow-Core/OSWindowAttributes.class.st:296]: For following iteration: format

**Review by guillep — APPROVED** [review]
Good to me, for the next iteration it would be nice to reformat some code and add some comments!

**Review by jordanmontt — APPROVED** [review]
Nice :)

**Comment by MarcusDenker** [comment]
Failing tests:

- StDebuggerTest>>testCopyStackToClipboard
- StDebuggerActionModelTest>>testCopyStackToClipboard
- ProperMethodCategorizationTest>>testInstanceSideInitializeMethodNeedsToBeInInitializationProtocol
- ProperMethodCategorizationTest>>testTearDownMethodInSUnitTestsNeedsToBeInRunningProtocol
- RubSmalltalkEditorTest>>testCopySelection

File: src/Morphic-Base/MorphicWindowManager.class.st (hunk 13 above)
Comment (RemiDufloer) [reply src/Morphic-Base/MorphicWindowManager.class.st:87]: they do nothing, I add the comment on the method

**Comment by guillep** [comment]
This is now green! Thanks @RemiDufloer !

**Pull request merged**